bioio/writers/ome_zarr_writer_2.py

```diff
@@ -0,0 +1,688 @@
+    more resolution in XY than in Z so this is a reasonable assumption.
+
+    The first shape returned will always be (1,1,1,shapes[0][3],shapes[0][4])
+    and the following will be a scaled number of slices scaled by the same
+    factor as the successive shapes.
+
+    This is an attempt to keep the total size of chunks the same across all
+    levels, by increasing the number of slices for downsampled levels.
+    This is making a basic assumption that each of the shapes is a downsampled
+    version of the previous shape.
+
+    For example, in a typical case, if the second level is scaled down by 1/2
+    in X and Y, then the second chunk size will have 4x the number of slices.
+
+    Parameters
+    ----------
+    shapes:
+        List of all multiresolution level shapes
+
+    Returns
+    -------
+    List of chunk sizes for per-slice chunking
+    """
+    # assumes TCZYX order
+    shape0 = shapes[0]
+    chunk_sizes = []
+    # assume starting with single slice
+    chunk_sizes.append((1, 1, 1, shape0[3], shape0[4]))
+    for i in range(1, len(shapes)):
+        last_chunk_size = chunk_sizes[i - 1]
+        scale = get_scale_ratio(shapes[i - 1], shapes[i])
+        # assumes that scale ratio for TCZ is 1 and only downsampled in xy.
+        chunk_sizes.append(
+            (
+                1,
+                1,
+                (int(scale[4] * scale[3] * last_chunk_size[2])),
+                int(last_chunk_size[3] / scale[3]),
+                int(last_chunk_size[4] / scale[4]),
+            )
+        )
+    return chunk_sizes
+
+
+class OmeZarrWriter:
+    """
+    Class to write OME-Zarr files.
+    Example usage:
+
+    .. code-block:: python
+
+            from ome_zarr_writer import
+                OmeZarrWriter,
+                compute_level_shapes,
+                compute_level_chunk_sizes_zslice
+
+            # We need to compute the shapes and chunk sizes for each
+            # desired multiresolution level.
+            shapes = compute_level_shapes(input_shape, scaling, num_levels)
+            chunk_sizes = compute_level_chunk_sizes_zslice(shapes)
+
+            # Create an OmeZarrWriter object
+            writer = OmeZarrWriter()
+
+            # Initialize the store. Use s3 url or local directory path!
+            writer.init_store(str(save_uri), shapes, chunk_sizes, im.dtype)
+
+            # Write the image.
+            # This will compute downsampled levels on the fly.
+            # Adjust t batch size based on dask compute capacity.
+            writer.write_t_batches_array(im, tbatch=4)
+
+            # Generate a metadata dict and write it to the zarr.
+            meta = writer.generate_metadata(
+                image_name="my_image_name",
+                channel_names=my_channel_names,
+                physical_dims=physical_scale,
+                physical_units=physical_units,
+                channel_colors=my_channel_colors,
+            )
+            writer.write_metadata(meta)
+    """
```

Comment:
the above example is the typical usage/code flow of how to use this writer

Reply:
Does this also resolve the desire to be able to append new data to the zarr store over time? I.e. while a microscope is taking a timelapse, can we store each timepoint as they come in?

If its not possible, no worries, if it is possible, I would say that is worth highlighting

Reply:
This code isn't made for modifying an existing one, I believe it's only been tested for when you are ready to write everything all at once... the use case of converting pre-existing files in other formats.  But it might work in the scenario you describe!